File: 08-04-2024/Optimal Strategy For A Game.py
#User function Template for python3


#Function to find the maximum possible amount of money we can win.
class Solution:
    def optimalStrategyOfGame(self,n, arr):
        # Your Code Start
        dp = [[0] * n for _ in range(n)]

        for g in range(n):
            for i in range(n - g):
                j = i + g
                if g == 0:
                    dp[i][j] = arr[i]
                elif g == 1:
                    dp[i][j] = max(arr[i], arr[j])
                else:
                    dp[i][j] = max(
                        arr[i] + min(dp[i + 2][j], dp[i + 1][j - 1]),
                        arr[j] + min(dp[i + 1][j - 1], dp[i][j - 2])
                    )

        return dp[0][n - 1]
        #Your Code End




#{ 
 # Driver Code Starts
#Initial Template for Python 3
import atexit
import io
import sys

# Contributed by : Nagendra Jha

if __name__ == '__main__':
    test_cases = int(input())
    for cases in range(test_cases):
        n = int(input())
        arr = list(map(int,input().strip().split()))
        ob = Solution()
        print(ob.optimalStrategyOfGame(n,arr))

# } Driver Code Ends

File: 08-04-2024/Optimal Strategy For A Game.java
//{ Driver Code Starts
import java.util.*;
import java.io.*;
import java.lang.*;

class OptimalStrategy
{
    public static void main (String[] args) {
        
        //taking input using Scanner class
        Scanner sc = new Scanner(System.in);
        
        //taking total number of testcases
        int t = sc.nextInt();
        
        while(t-- > 0)
        {
            //taking number of elements
            int n = sc.nextInt();
            int arr[] = new int[n];
            
            //inserting the elements
            for(int i = 0; i < n; i++)
                arr[i] = sc.nextInt();
                
           //calling the countMaximum() method of class solve
           System.out.println(new solve().countMaximum(n, arr)); 
        }
    }
    
    
}
// } Driver Code Ends





class solve
{
    //Function to find the maximum possible amount of money we can win.
    static long countMaximum(int n, int arr[])
    {
        //Your Code Start
            int[][] dp=new int[n][n];
        
        for(int g=0; g<n; g++) {
            for(int i=0, j=g; j<n; i++, j++){
                if(g==0) dp[i][j]=arr[i];
                else if(g==1) dp[i][j]=Math.max(arr[i], arr[j]);
                else{
                    dp[i][j]=Math.max(
                        arr[i]+Math.min(dp[i+2][j], dp[i+1][j-1]),
                        arr[j]+Math.min(dp[i+1][j-1], dp[i][j-2])
                    );
                }
            }
        }
        
        return dp[0][n-1];
        // Your code End
    }
}
